Mathlib/GroupTheory/ClassEquation.lean

```diff
@@ -0,0 +1,76 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Eric Rodriguez
+-/
+
+import Mathlib.GroupTheory.GroupAction.Quotient
+import Mathlib.Algebra.BigOperators.Finprod
+import Mathlib.Data.Finite.Card
+import Mathlib.Algebra.Group.ConjFinite
+
+/-!
+# Class Equation
+
+This file establishes the class equation for finite groups.
+
+## Main statements
+
+* `Group.card_center_add_sum_card_noncenter_eq_card`: The **class equation** for finite groups.
+  The cardinality of a group is equal to the size of its center plus the sum of the size of all its
+  nontrivial conjugacy classes.
+
+-/
+
+open ConjAct MulAction ConjClasses
+
+open scoped BigOperators
+
+/-- The conjugacy classes form a partition of G, in terms of cardinality. A triviality but helps
+to prove the class equation. -/
+theorem sum_conjClasses_card_eq_card [Group G] [Fintype (ConjClasses G)]
+    [∀ x : ConjClasses G, Fintype x.carrier] [Fintype G] :
+    ∑ x : ConjClasses G, x.carrier.toFinset.card = Fintype.card G := by
+  let e : Quotient (orbitRel (ConjAct G) G) ≃ ConjClasses G :=
+    Quotient.congrRight fun g h ↦ mem_orbit_conjAct
+  classical
+  rw [← e.sum_comp, card_eq_sum_card_group_div_card_stabilizer (ConjAct G) G]
```

Comment:
I feel like there should be a more direct proof that doesn't involve stabilizer subgroups. Maybe prove a bijection of the form `G ≃ ∑ x : ConjClasses G, x.carrier`?

Reply:
Ah, maybe using `selfEquivSigmaOrbits`?

Reply:
I tried and I'm just ending up fighting against the API because they're technically two different quotient types, but I can't `subst` anything in to stop making a mess (just copy+paste this into the file):

```lean4
example [Group G] [Fintype (ConjClasses G)]
    [∀ x : ConjClasses G, Fintype x.carrier] [Fintype G] :
    ∑ x : ConjClasses G, x.carrier.toFinset.card = Fintype.card G := by
  suffices : (Σ x : ConjClasses G, x.carrier) ≃ G
  · simpa using Fintype.card_congr this
  refine Equiv.trans ?_ (selfEquivSigmaOrbits (ConjAct G) G).symm
  refine Equiv.sigmaCongr ?_ ?_
  · rw [ConjClasses, orbitRel.Quotient]
    suffices : IsConj.setoid G = orbitRel (ConjAct G) G
    · rw [this]
    ext
    rw [orbitRel_conjAct]
    rfl
  sorry
```

If you have any ideas I'm all ears

Reply:
Does this help?
```
def equivSigmaConjClasses : G ≃ Σ x : ConjClasses G, x.carrier := by
  simp_rw [carrier_eq_preimage_mk]
  exact (Equiv.sigmaFiberEquiv ConjClasses.mk).symm
```

Reply:
ahh, awesome, I don't know what I was overcomplicating everything!